File: TransCoder_tokenized_test_set_functions/python/EXPONENTIAL_SQUARING_FAST_MODULO_MULTIPLICATION.py
def exponentiation ( bas, exp ) :
    if ( exp == 0 ) :
        return 1;
        
    if ( exp == 1 ) :
        return bas % N;
        
    t = exponentiation ( bas, int ( exp / 2 ) );
    
    t = ( t * t ) % N;
    
    if ( exp % 2 == 0 ) :
        return t;
        
    else :
        return ( ( bas % N ) * t ) % N;
        
    

File: TransCoder_tokenized_test_set_functions/cpp/EXPONENTIAL_SQUARING_FAST_MODULO_MULTIPLICATION.cpp
long int exponentiation ( long int base, long int exp ) {
  if ( exp == 0 ) return 1;
  if ( exp == 1 ) return base % N;
  long int t = exponentiation ( base, exp / 2 );
  t = ( t * t ) % N;
  if ( exp % 2 == 0 ) return t;
  else return ( ( base % N ) * t ) % N;
}